### PR TITLE
[CPU] Linearize gpt_oss model and add example to quantize it to w4a8

### PR DESCRIPTION
**Pull request by isharif168**
SUMMARY:
Add code to convert the gpt_oss model and then quantize model to w8a8

TEST PLAN:
Tested with gpt_oss model and run using vllm

Sample output

=== Prompt 0 ===
<|start|>system<|message|>You are ChatGPT, a large language model trained by OpenAI.
Knowledge cutoff: 2024-06
Current date: 2025-12-04

Reasoning: medium

Valid channels: analysis, commentary, final. Channel must be included for every message.<|end|><|start|>developer<|message|># Instructions

You are a helpful assistant.

<|end|><|start|>user<|message|>Give 3 reasons to use AI.<|end|><|start|>assistant

--- Candidate 0 ---
analysisWe need to answer with three reasons to use AI. Should be short. Provide 3 reasons. Let's do it.assistantcommentaryHere are three reasons to use AI:

Personalized learning – AI can adapt to your learning style and pace, providing customized lessons that suit your learning style and progress.
Skill development – AI can help you develop skills that are in demand, such as coding, data analysis, and data science.
Supportive design
We should provide a short answer.assistantfinalHere are three reasons to use AI:

Personalized learning – AI adapts to your learning style and pace, giving you personalized lessons that fit your needs.
Skill development – AI helps you learn and practice skills that are in demand, like coding, data analysis, and data‑science skills.
Real‑world impact – AI helps you practice real-world skills in real life, such as learning new skills and applying them in real life.
finish_reason: stop
num_tokens: 226

### CODE REVIEW

**Comment by dsikka** [comment]
Thank you! Do you have a sample model you can share with us?

**Comment by isharif168** [comment]
Hi @dsikka , I have added the example as well which will convert and quantize the gpt_oss_20b mode
`python gpt_oss_20b_example.py`
This should download the model and also quantize it

**Comment by dsikka** [comment]
Thank you! I will take a look through this example this week

**Review by dsikka — APPROVED** [review]
Thank you for your contribution! This looks good to me. Some comments and suggestions. 

1. Did you mean to quantize this to W4A8 or W8A8?  
2. For the replacement code that you're applying, we have a class which registers the modules and replaces them for you: https://github.com/vllm-project/llm-compressor/blob/5f6c8dbca1ce50f92a70560d23a675a310ce17ad/src/llmcompressor/modeling/moe_context.py#L29. We should consider applying this in a follow up.
3. I would make a note that you're working on vLLM support and link the PR so that others are aware how to run the checkpoint
4. You're using the QuantizationModifier but Int4 Quantization does better with GPTQ and AWQ. Consider using ones of the modifiers for better accuracy

I made small updates to your script which I've written out here which I suggest adding:

```python 
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer

from llmcompressor import oneshot
from llmcompressor.modifiers.quantization import QuantizationModifier

from compressed_tensors.quantization import QuantizationScheme
from compressed_tensors.quantization.quant_args import (
    QuantizationArgs,
    QuantizationStrategy,
    QuantizationType,
)
from llmcompressor.utils import dispatch_for_generation

from llmcompressor.modeling.gpt_oss import convert_model_for_quantization_gptoss


def main():
    MODEL_ID = "openai/gpt-oss-20b"
    BASE_NAME = MODEL_ID.rstrip("/").split("/")[-1]
    OUTPUT_DIR = f"{BASE_NAME}-w4a8-channelwise"

    print(f"[GPT-OSS] Loading model: {MODEL_ID}")
    model = AutoModelForCausalLM.from_pretrained(
        MODEL_ID,
        torch_dtype=torch.bfloat16,
        trust_remote_code=True,
    )
    tokenizer = AutoTokenizer.from_pretrained(MODEL_ID, trust_remote_code=True)

    # ---- GPT-OSS MoE → linear experts conversion ----
    print("[GPT-OSS] Converting fused MoE experts to LinearExperts for quantization...")
    convert_model_for_quantization_gptoss(model)
    print("[GPT-OSS] Conversion completed.")

    # ---- Quantization config: W4A8 (int4 weights, int8 activations) ----

    # Weights: 4-bit, channelwise, symmetric, static
    weights_args = QuantizationArgs(
        num_bits=4,
        type=QuantizationType.INT,
        strategy=QuantizationStrategy.CHANNEL,
        symmetric=True,
        dynamic=False,
    )

    # Activations: 8-bit, per-token, asymmetric, dynamic
    activations_args = QuantizationArgs(
        num_bits=8,
        type=QuantizationType.INT,
        strategy=QuantizationStrategy.TOKEN,
        symmetric=False,
        dynamic=True,
        observer=None,
    )

    # Apply to all Linear layers, excluding lm_head
    scheme = QuantizationScheme(
        targets=["Linear"],
        weights=weights_args,
        input_activations=activations_args,
    )

    recipe = QuantizationModifier(
        config_groups={"group_0": scheme},
        ignore=["lm_head"],
    )

    print(f"[GPT-OSS] Starting oneshot quantization → {OUTPUT_DIR}")
    oneshot(
        model=model,
        recipe=recipe,
        tokenizer=tokenizer,
        trust_remote_code_model=True,
    )
    print("\n\n")
    print("========== SAMPLE GENERATION ==============")
    dispatch_for_generation(model)
    input_ids = tokenizer("Hello my name is", return_tensors="pt").input_ids.to(
        model.device
    )
    output = model.generate(input_ids, max_new_tokens=100)
    print(tokenizer.decode(output[0]))
    print("==========================================\n\n")

    print(f"[GPT-OSS] Quantization finished. Quantized model being written to: {OUTPUT_DIR}")
    model.save_pretrained(OUTPUT_DIR, save_compressed=True)
    tokenizer.save_pretrained(OUTPUT_DIR)


if __name__ == "__main__":
    main()

```

File: examples/quantization_w4a8/gpt_oss_20b_example.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    MODEL_ID = "openai/gpt-oss-20b"
+    BASE_NAME = MODEL_ID.rstrip("/").split("/")[-1]
+    OUTPUT_DIR = f"{BASE_NAME}-w4a8-channelwise"
+
+    print(f"[GPT-OSS] Loading model: {MODEL_ID}")
+    model = AutoModelForCausalLM.from_pretrained(
+        MODEL_ID,
+        torch_dtype=torch.bfloat16,
+        device_map="auto",
+        trust_remote_code=True,
```
Comment (kylesayrs) [inline examples/quantization_w4a8/gpt_oss_20b_example.py:25]: GPT OSS does not contain remote code, so this flag should not be required.

File: src/llmcompressor/modeling/gpt_oss.py [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+        out = torch.zeros_like(x)
+
+        # GPT-OSS router uses an extra "no expert" bucket at index E
+        with torch.no_grad():
+            expert_mask = torch.nn.functional.one_hot(
+                router_indices, num_classes=num_experts_plus_dummy
+            ).permute(2, 1, 0)
+            expert_hit = torch.greater(expert_mask.sum(dim=(-1, -2)), 0).nonzero()
+
+        for idx in expert_hit:
```
Comment (kylesayrs) [inline src/llmcompressor/modeling/gpt_oss.py:152]: In the future, we'd like to expand this to support the `calibrate_all_experts` option. However for this example, the forward function is not used anyways, so this is fine.

File: src/llmcompressor/modeling/gpt_oss.py [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+class ExpertMeta:
+    path: str
+    hidden_size: int
+    intermediate_size: int
+    num_experts: int
+    device: torch.device
+    dtype: torch.dtype
+
+
+def get_module_by_path(root: nn.Module, dotpath: str) -> nn.Module:
```
Comment (kylesayrs) [inline src/llmcompressor/modeling/gpt_oss.py:180]: `torch.nn.Module.get_submodule()`?

File: src/llmcompressor/modeling/gpt_oss.py [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+    m: nn.Module = root
+    if not dotpath:
+        return root
+    for p in dotpath.split("."):
+        m = getattr(m, p)
+    return m
+
+
+def set_module_by_path(root: nn.Module, dotpath: str, new_module: nn.Module) -> None:
```
Comment (kylesayrs) [inline src/llmcompressor/modeling/gpt_oss.py:189]: `torch.nn.Module.set_submodule()`?

File: src/llmcompressor/modeling/gpt_oss.py [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+
+def find_experts(model: nn.Module) -> List[ExpertMeta]:
+    """
+    Locate GPT-OSS MoE expert modules under model.model.layers[*].mlp.experts.
+    """
+    metas: List[ExpertMeta] = []
+    for li, layer in enumerate(model.model.layers):
+        experts = layer.mlp.experts
+        device = next(experts.parameters(), torch.zeros(())).device
+        dtype = next(experts.parameters(), torch.zeros(())).dtype
```
Comment (kylesayrs) [inline src/llmcompressor/modeling/gpt_oss.py:203]: I think that we should error if the experts ever don't have parameters, not default to cpu

File: src/llmcompressor/modeling/gpt_oss.py [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+    parts = dotpath.split(".")
+    parent = get_module_by_path(root, ".".join(parts[:-1]))
+    setattr(parent, parts[-1], new_module)
+
+
+def find_experts(model: nn.Module) -> List[ExpertMeta]:
+    """
+    Locate GPT-OSS MoE expert modules under model.model.layers[*].mlp.experts.
+    """
+    metas: List[ExpertMeta] = []
```
Comment (kylesayrs) [inline src/llmcompressor/modeling/gpt_oss.py:199]: In the future, I think we can avoid using a meta data class and instead just convert by accessing the relevant data directly

**Comment by kylesayrs** [comment]
@isharif168 @dsikka I've created an issue to address any follow up we may want to add, but this is good to land as is now!

https://github.com/vllm-project/llm-compressor/issues/2159

**Comment by isharif168** [comment]
Thanks @isharif168 and @kylesayrs for the reviews and comments.
Here is the corresponding vllm change that will support w4a8 for gpt_oss
https://github.com/vllm-project/vllm/pull/29315

Thanks.
